File: android/app/src/main/java/com/mewo/reader/data/HostedSessionStore.kt
package com.mewo.reader.data

import android.content.Context
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

data class HostedSession(
    val baseUrl: String = "",
    val username: String? = null,
    val token: String? = null,
) {
    val signedIn: Boolean
        get() = !token.isNullOrBlank() && normalizedUrl() != null

    fun normalizedUrl(): String? {
        val url = normalizeBaseUrl(baseUrl)
        return url.ifBlank { null }
    }
}

fun normalizeBaseUrl(raw: String): String {
    var url = raw.trim().trimEnd('/')
    if (url.isEmpty()) return ""
    if (!url.startsWith("http://") && !url.startsWith("https://")) {
        url = "http://$url"
    }
    return url
}

class HostedSessionStore(context: Context) {
    private val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    private val _session = MutableStateFlow(read())
    val session: StateFlow<HostedSession> = _session.asStateFlow()

    fun setBaseUrl(url: String) {
        prefs.edit().putString(KEY_URL, url).apply()
        _session.update { it.copy(baseUrl = url) }
    }

    fun setSignedIn(username: String, token: String) {
        prefs.edit()
            .putString(KEY_USER, username)
            .putString(KEY_TOKEN, token)
            .apply()
        _session.update { it.copy(username = username, token = token) }
    }

    fun signOut() {
        prefs.edit()
            .remove(KEY_USER)
            .remove(KEY_TOKEN)
            .apply()
        _session.update { it.copy(username = null, token = null) }
    }

    private fun read(): HostedSession = HostedSession(
        baseUrl = prefs.getString(KEY_URL, "").orEmpty(),
        username = prefs.getString(KEY_USER, null),
        token = prefs.getString(KEY_TOKEN, null),
    )

    companion object {
        private const val PREFS = "mewo_hosted"
        private const val KEY_URL = "base_url"
        private const val KEY_USER = "username"
        private const val KEY_TOKEN = "token"
    }
}

File: ios/Mewo/Data/HostedSessionStore.swift
import Foundation
import Observation

struct HostedSession: Equatable, Sendable {
    var baseUrl: String = ""
    var username: String?
    var token: String?

    var signedIn: Bool {
        !(token ?? "").isEmpty && normalizedUrl() != nil
    }

    func normalizedUrl() -> String? {
        let url = normalizeBaseUrl(baseUrl)
        return url.isEmpty ? nil : url
    }
}

func normalizeBaseUrl(_ raw: String) -> String {
    var url = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    while url.hasSuffix("/") { url.removeLast() }
    if url.isEmpty { return "" }
    if !url.hasPrefix("http://") && !url.hasPrefix("https://") {
        url = "http://\(url)"
    }
    return url
}

@Observable
final class HostedSessionStore {
    private let defaults = UserDefaults(suiteName: "mewo_hosted") ?? .standard
    private(set) var session: HostedSession

    init() {
        session = HostedSession(
            baseUrl: defaults.string(forKey: "base_url") ?? "",
            username: defaults.string(forKey: "username"),
            token: Keychain.readToken()
        )
    }

    func setBaseUrl(_ url: String) {
        defaults.set(url, forKey: "base_url")
        session.baseUrl = url
    }

    func setSignedIn(username: String, token: String) {
        defaults.set(username, forKey: "username")
        Keychain.saveToken(token)
        session.username = username
        session.token = token
    }

    func signOut() {
        defaults.removeObject(forKey: "username")
        Keychain.deleteToken()
        session.username = nil
        session.token = nil
    }
}
